[feature] Add -CustomPipeName to `pwsh` and `Enter-PSHostProcess`

TylerLeonhardt opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->  

## PR Summary

This allows a user to start PowerShell up with the name of the named pipe that is used for cross process communication (I.e. `Enter-PSHostProcess`).

### Example:

terminal 1
```
pwsh -custompipename mypipe
```
terminal 2
```
Enter-PSHostProcess -CustomPipeName mypipe
```

## PR Context  

Today, it is impossible to predictively connect to a PowerShell process without knowing its process id. If you're stopping and starting a powershell process often... you have no mechanism for figuring out the process id.

This introduces a consistent way to connect to a PowerShell process by specifying the NamedPipe yourself.

This is very nice in PowerShell hosted scenarios - for example, in Azure Functions where PowerShell is hosted. This enables the ability to consistently connect to the PowerShell instance hosted there without any additional information besides the CustomPipeName that can be reused.

This feature also means that in the PowerShell extension for VSCode, we can achieve "Attach to Process" debugging with a single stroke of the <kbd>F5</kbd> key.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.  
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**  
    - [x] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed: https://github.com/MicrosoftDocs/PowerShell-Docs/issues/3755
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

TylerLeonhardt committed:
Add DebugPipeName capability

TylerLeonhardt committed:
Add DebugPipeName capability

TylerLeonhardt committed:
[feature] minor fix for Windows

TylerLeonhardt committed:
[feature] minor fix for Windows

TylerLeonhardt committed:
[feature] misc code factor fixes

TylerLeonhardt committed:
[feature] misc code factor fixes

TylerLeonhardt committed:
[feature] add more tests and pass command down to ServerMode

TylerLeonhardt committed:
[feature] add more tests and pass command down to ServerMode

TylerLeonhardt committed:
[feature] fix path for Windows

TylerLeonhardt committed:
[feature] fix path for Windows

TylerLeonhardt committed:
[feature] add erroraction stop and rely on a better way to wait

TylerLeonhardt committed:
[feature] add erroraction stop and rely on a better way to wait

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -773,6 +782,18 @@ private void ParseHelper(string[] args)
+                        break;
+                    }
+
+                    _debugPipeName = args[i];
```
There's a max length allowed for the pipename (I hit this on macOS when enabling this).  I have it as a comment in the code somewhere.  We should validate here otherwise it'll fail when creating the pipe.

TylerLeonhardt replied:
Oh right yes. I forgot about that. It's 104 on macOS, 108 on Linux. I'll address that

TylerLeonhardt replied:
fixed.

PaulHigin commented on src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx:
```diff
@@ -186,6 +186,9 @@
     <value>Cannot process the command because -Configuration requires an argument that is a remote endpoint configuration name. Specify this argument and try again.</value>
   </data>
+  <data name="MissingDebugPipeNameArgument" xml:space="preserve">
+    <value>Cannot process the command because -DebugPipeName requires an argument that is a name the pipe you want to use. Specify this argument and try again.</value>
```
`...that is a name **of** the pipe you want to use. ...`

PaulHigin commented on src/System.Management.Automation/engine/hostifaces/LocalConnection.cs:
```diff
@@ -47,6 +47,10 @@ internal sealed partial class LocalRunspace : RunspaceBase
         internal LocalRunspace(PSHost host, InitialSessionState initialSessionState, bool suppressClone)
             : base(host, initialSessionState, suppressClone)
         {
+            if (RemoteSessionNamedPipeServer.IPCNamedPipeServer == null)
```
I am worried about this because `CreateIPCNamedPipeServerSingleton` is called in a static constructor of RemoteSessionNamedPipeServer, so this creates a potential race condition.  I think `CreateIPCNamedPipeServerSingleton` will need to **force** the creation of a server object with the provided pipe name, in case the static constructor is run first.

Also, we don't need the null check.

TylerLeonhardt replied:
I removed it from the static constructor and moved it here so that I could access the `InitialSessionState` of `LocalRunspace`.

PaulHigin replied:
I feel this is the wrong approach.  This default IPC listener should not be tied to the creation of a local runspace, and should remain as it is now.  I think it is fine to start the powershell.exe host with a switch specifying a custom pipe name for special applications like PowerShell VSCode extension, but it should either replace the default IPC or run as an optional second IPC channel.

TylerLeonhardt replied:
I see what you're saying. 

The problem I faced was that I didn't see the clearest route for grabbing the pipe name from the pwsh.exe parameters to the static constructor.

Going through the `initialsessionstate`, to the runspace seemed like the best option because it would be easily exposed in the PowerShell SDK. You could set the `PipeName` in the `InitialSessionState` and get this feature.

Whats your recommendation for implementing this such that it could also be easily exposed in the SDK as well as via `pwsh.exe`?

PaulHigin reviewed (changes requested):
I will finish the review later, but have a couple of comments now.
Thanks!

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Core/Enter-PSHostProcess.Tests.ps1:
```diff
@@ -2,54 +2,106 @@
-            Remove-Item $powershell_started -Force -ErrorAction SilentlyContinue
+            "enter-pshostprocess -id $($pwsh.Id) -erroraction stop
+            `$pid
+            exit-pshostprocess" | pwsh -c - | Should -Be $pwsh.Id
```
Could you please use right case for commends in the string.  `Enter-PSHostProcess -Id $($pwsh.Id) -ErrorAction Stop`

Below too.

TylerLeonhardt replied:
I'm going to blame this one on @SteveL-MSFT 😜 

Will fix.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Core/Enter-PSHostProcess.Tests.ps1:
```diff
@@ -2,54 +2,106 @@
-        Wait-UntilTrue { Test-Path $powershell_started }
+    Context "By DebugPipeName" {
+        BeforeAll {
+            $pipeName = [System.IO.Path]::GetRandomFileName()
```
Can we use New-TemporaryFile?

PaulHigin commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -829,10 +839,11 @@ private void ProcessListeningThread(object state)
         /// <param name="configurationName">Name of the configuration to use.</param>
-        internal static void RunServerMode(string configurationName)
+        /// <param name="debugPipeName">Name of the pipe to use if it was specified at startup.</param>
+        internal static void RunServerMode(string configurationName, string debugPipeName)
```
PowerShell Direct should never need to create a custom named IPC pipe, passing in `debugPipeName` is unnecessary.

PaulHigin commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -858,7 +869,8 @@ internal static void RunServerMode(string configurationName)
         /// starts the client listening thread.
         /// </summary>
-        internal static void CreateIPCNamedPipeServerSingleton()
+        /// <param name="debugPipeName">Name of the pipe name to use.</param>
```
`debugPipeName` name presumes IPC is for debugging, but that is not necessarily the case.  I think we should use a different name like `customPipeName` or `IPCPipeName`.

PaulHigin reviewed (changes requested):
Let's remove the connection of pipe listener creation to local runspace, and think about whether we want to replace the existing default IPC channel or create a second one when starting powershell.exe in `DebugPipeName` mode, although I think we should use a different name for the switch, like `IPCPipe` or `CustomIPCPipe`, or something.

TylerLeonhardt committed:
refactor to a static public method

TylerLeonhardt committed:
refactor to a static public method

TylerLeonhardt committed:
[feature] change test casing

TylerLeonhardt committed:
[feature] change test casing

TylerLeonhardt commented:
@PaulHigin I've refactored this quite a bit based on your offline feedback. The same experience is there for a user at the CLI.

For users that are going through the API, the can use:

```csharp
S.M.A.Remoting.RemoteSessionNamedPipeServer.CreateCustomNamedPipeServer("mypipe");
```

There are 2 things left to do here (which is why I moved to WIP):

- change the name from `DebugPipeName` to `CustomPipeName` - everyone feel free to comment

- figure out what should happen when this static function is called multiple times - Should it throw? Or should it overwrite?

TylerLeonhardt committed:
dispose of the server if we recall the function

TylerLeonhardt committed:
dispose of the server if we recall the function

PaulHigin commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -429,14 +430,19 @@ internal static int NamedPipeBufferSizeForRemoting
                 System.Diagnostics.Process.GetCurrentProcess(), appDomainName));
         }
 
+        internal static RemoteSessionNamedPipeServer CreateRemoteSessionNamedPipeServer(string pipeName)
```
This seems unnecessary, since it just wraps a constructor taking the same argument.

TylerLeonhardt replied:
refactored.

PaulHigin commented on src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs:
```diff
@@ -1770,6 +1779,28 @@ public NamedPipeConnectionInfo()
+            string debugPipeName,
+            int openTimeout)
+        {
+            DebugPipeName = debugPipeName;
```
We should validate that `debugPipeName` parameter is not null or empty.

TylerLeonhardt replied:
added.

daxian-dbw commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -91,6 +92,17 @@ public PSHostProcessInfo HostProcessInfo
+        /// Gets and sets the Named Pipe name to connect to.
+        /// </summary>
+        [Parameter(Position = 0, Mandatory = true, ValueFromPipeline = true, ParameterSetName = EnterPSHostProcessCommand.PipeNameParameterSet)]
+        [ValidateNotNull()]
```
`ValidateNotNull` is not needed. `Mandatory = true` implies no-null and no-empty input.

TylerLeonhardt replied:
removed.

daxian-dbw commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -429,14 +430,19 @@ internal static int NamedPipeBufferSizeForRemoting
             string appDomainName = NamedPipeUtils.GetCurrentAppDomainName();
 
-            return new RemoteSessionNamedPipeServer(NamedPipeUtils.CreateProcessPipeName(
+            return CreateRemoteSessionNamedPipeServer(NamedPipeUtils.CreateProcessPipeName(
```
It's better to revert this change. It adds one additional method call without much value.

TylerLeonhardt replied:
fixed.

daxian-dbw commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -203,16 +235,19 @@ private Runspace CreateNamedPipeRunspace(int procId, string appDomainName)
+                            StringUtil.Format(
+                                RemotingErrorIdStrings.EnterPSHostProcessCannotConnectToProcess,
+                                msgAppDomainName,
+                                connectionInfo.ProcessId,
```
`ProcessId` property would be `0` by default when using `-DebugPipeName`. The error message may need to be revisited.

TylerLeonhardt replied:
added an error message for debugpipename scenarios

TylerLeonhardt committed:
address all feedback

TylerLeonhardt committed:
address all feedback

TylerLeonhardt commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -341,6 +342,14 @@ internal bool ShowVersion
+        {
+            get
+            {
+                return _debugPipeName;
```
I was just following the rest of the properties. If you'd like me to change it, I can.

PaulHigin replied:
I feel consistency is good.  We can address code clean up in a different issue.

TylerLeonhardt committed:
[feature] add test for API and help text

TylerLeonhardt committed:
[feature] add test for API and help text

TylerLeonhardt commented on src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx:
```diff
@@ -186,6 +186,12 @@
+    <value>Cannot process the command because -DebugPipeName requires an argument that is a name of the pipe you want to use. Specify this argument and try again.</value>
+  </data>
+  <data name="DebugPipeNameTooLong" xml:space="preserve">
+    <value>Cannot process the command because -DebugPipeName specified is too long.</value>
```
addressed.

TylerLeonhardt commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -91,6 +92,17 @@ public PSHostProcessInfo HostProcessInfo
+        /// <summary>
+        /// Gets and sets the Named Pipe name to connect to.
+        /// </summary>
+        [Parameter(Position = 0, Mandatory = true, ValueFromPipeline = true, ParameterSetName = EnterPSHostProcessCommand.PipeNameParameterSet)]
```
fixed.

TylerLeonhardt commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -189,7 +189,8 @@ internal class CommandLineParameterParser
             "workingdirectory",
-            "removeworkingdirectorytrailingcharacter"
+            "removeworkingdirectorytrailingcharacter",
+            "debugpipename"
```
added.

TylerLeonhardt commented:
Addressed all feedback 👍

PaulHigin commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -552,6 +553,7 @@ static RemoteSessionNamedPipeServer()
 
             CreateIPCNamedPipeServerSingleton();
+
 #if !CORECLR // There is only one AppDomain per application in CoreCLR, which would be the default
```
This code attempts to dispose the default pipe server on process (appdomain) exit, but it is disabled for CORECLR and so is dead code.  This code should be removed or fixed to dispose both the default and custom pipe server objects.  Can you create a new Issue to look into it?

TylerLeonhardt replied:
Create the issue:
https://github.com/PowerShell/PowerShell/issues/8925

PaulHigin reviewed (approved):
LGTM.  But please create new Issue about disposing pipe server objects on process exit (please see comment).

daxian-dbw commented on src/Microsoft.PowerShell.ConsoleHost/resources/CommandLineParameterParserStrings.resx:
```diff
@@ -186,6 +186,15 @@
+    <value>Cannot process the command because -DebugPipeName requires an argument that is a name of the pipe you want to use. Specify this argument and try again.</value>
+  </data>
+  <data name="DebugPipeNameTooLongOnLinux" xml:space="preserve">
+    <value>Cannot process the command because -DebugPipeName specified is too long. Pipe paths on Linux can be up to 108 characters long. Your pipe path: {0}</value>
```
> Pipe paths on Linux can be up to 108 characters long. Your pipe path: {0}

What the user specifies is the pipe name, not the pipe path, which would be `TEMP/CoreFxPipe_xxx...`. So (1) better say "pipe name" instead of "pipe path"; (2) is the 108 the length limit for the path or the name? better make sure we give the length limit for the name.

TylerLeonhardt replied:
cleaned up!
```
pwsh -debugpipename sdfhsardfghghjsdfdghkjhsdfghsdfghdfgssdfgsdfhkjsdhkjsdhksdfghk
Cannot process the command because -DebugPipeName specified is too long. Pipe names on this platform can be up to 55 characters long. Your pipe name 'sdfhsardfghghjsdfdghkjhsdfghsdfghdfgssdfgsdfhkjsdhkjsdhksdfghk' is 62 characters.
```

daxian-dbw reviewed (approved):
LGTM except for one comment about the resource string

TylerLeonhardt committed:
[feature] address dongbo's feedback

TylerLeonhardt committed:
[feature] address dongbo's feedback

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -773,6 +782,33 @@ private void ParseHelper(string[] args)
+
+                    if (!Platform.IsWindows)
+                    {
+                        int maxNameLength = (Platform.IsLinux ? 108 : 104) - Path.GetTempPath().Length;
```
Prefer consts for the max length values

TylerLeonhardt replied:
added.

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -773,6 +782,33 @@ private void ParseHelper(string[] args)
+                        int maxNameLength = (Platform.IsLinux ? 108 : 104) - Path.GetTempPath().Length;
+                        if (args[i].Length > maxNameLength)
+                        {
+                            WriteCommandLineError(
```
Missing a test case for this

TylerLeonhardt replied:
added.

TylerLeonhardt committed:
[feature] add more tests and address Steve's feedback

TylerLeonhardt committed:
[feature] add more tests and address Steve's feedback

SteveL-MSFT commented on src/Microsoft.PowerShell.ConsoleHost/host/msh/CommandLineParameterParser.cs:
```diff
@@ -168,6 +168,9 @@ public override void WriteWarningLine(string message)
 
     internal class CommandLineParameterParser
     {
+        private const int _maxPipePathLengthLinux = 108;
```
I believe the convention for const is to use Pascal Casing.  This shows up as a style cop issue (I know the code base doesn't follow this consistently...)

```suggestion
        private const int MaxPipePathLengthLinux = 108;
```

TylerLeonhardt replied:
Fixed.

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Core/Enter-PSHostProcess.Tests.ps1:
```diff
@@ -2,54 +2,160 @@
+            $longPipeName = "DoggoipsumwaggywagssmolborkingdoggowithalongsnootforpatsdoingmeafrightenporgoYapperporgolongwatershoobcloudsbigolpupperlengthboy"
+
+            if (!$IsWindows) {
+                try {
```
Is there a reason you couldn't use the pattern:

{ ...scriptblock... | | Should -Throw -ErrorId "FullyQualifiedErrorId to validate"

TylerLeonhardt replied:
Ah! I was looking for that. Fixed.

TylerLeonhardt committed:
[feature] Steve's feedback

TylerLeonhardt committed:
[feature] Steve's feedback

TylerLeonhardt committed:
[feature] missed if !windows

TylerLeonhardt committed:
[feature] missed if !windows

TylerLeonhardt committed:
[feature] a command that will never fail

TylerLeonhardt committed:
[feature] a command that will never fail

TylerLeonhardt committed:
Merge branch 'debugpipename-enterpshostprocess' of github.com:TylerLeonhardt/PowerShell into debugpipename-enterpshostprocess

TylerLeonhardt committed:
[feature]switch to querying for the pid and remove unneed Wait-UntilTrue

TylerLeonhardt committed:
[feature] use exit code

TylerLeonhardt committed:
[feature] fix log message and Codacy and Code Factor fixes

TylerLeonhardt commented:
Ok I included a few CodeFactor and Codacy fixes as well. This should be good to go pending CI ✅

iSazonov commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -91,6 +92,16 @@ public PSHostProcessInfo HostProcessInfo
         }
 
+        /// <summary>
+        /// Gets or sets the Named Pipe name to connect to.
```
Should we mention "debug" here?

TylerLeonhardt replied:
fixed this with rename to CustomPipeName

iSazonov commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -91,6 +92,16 @@ public PSHostProcessInfo HostProcessInfo
+        /// <summary>
+        /// Gets or sets the Named Pipe name to connect to.
+        /// </summary>
+        [Parameter(Mandatory = true, ValueFromPipeline = true, ParameterSetName = EnterPSHostProcessCommand.PipeNameParameterSet)]
```
Why we need ValueFromPipeline ?

TylerLeonhardt replied:
removed.

iSazonov commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -343,7 +343,7 @@ public bool RestartListener
     /// thread loop.
     /// </summary>
-    internal sealed class RemoteSessionNamedPipeServer : IDisposable
+    public sealed class RemoteSessionNamedPipeServer : IDisposable
```
Why was it changed to public? It seems it is used only in tests - for tests we could use a test hook.

We need PowerShell Committee approval for new public API.

SteveL-MSFT replied:
This is needed for hosted PowerShell scenarios.  PS-Committee is not required to approve all public APIs, only if there is uncertainty or controversy.  As long as a PS-Maintainer is ok with the public API, we should not have the PS-Committee be a bottleneck.

iSazonov replied:
> This is needed for hosted PowerShell scenarios. 

For hosted scenarios I believe it'll be good to have xUnit tests.

TylerLeonhardt replied:
I've moved those select tests over to xUnit

iSazonov commented on src/System.Management.Automation/engine/remoting/common/RemoteSessionNamedPipe.cs:
```diff
@@ -600,14 +606,78 @@ public void Dispose()
+                    // Start the pipe server listening thread, and provide client connection callback.
+                    _customNamedPipeServer.StartListening(ClientConnectionCallback);
+                }
+                catch (Exception)
```
Can we catch more specific exception?

TylerLeonhardt replied:
There's actually quite a wide range of exceptions to catch here... and since we want this to be a best attempt so that powershell can still start if something goes wrong, it's best to catch just `Exception`.

iSazonov commented on src/System.Management.Automation/engine/remoting/common/RunspaceConnectionInfo.cs:
```diff
@@ -1722,20 +1722,29 @@ public string AppDomainName
         }
 
+        /// <summary>
+        /// Gets or sets the PipeName to connect to.
```
Should we mention "debug"?

TylerLeonhardt replied:
fixed this with rename to CustomPipeName

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Core/Enter-PSHostProcess.Tests.ps1:
```diff
@@ -2,54 +2,158 @@
+            $ps = [powershell]::Create()
+            $ps.Runspace = $rs
+            $ps.AddScript('$pid').Invoke() | Should -Be $pwsh.Id
+            $rs.Dispose()
```
Please put this in `finally` block (try {} finally {})

TylerLeonhardt replied:
fixed.

TylerLeonhardt committed:
Change DebugPipeName to CustomPipeName

TylerLeonhardt commented:
I've addressed @PaulHigin's final point which is that `DebugPipeName` is not totally accurate as this pipe could be used for other IPC scenarios. 

It's mostly used for debugging though... and our usecase is really for debugging... so I've called this out in the `-help` page.

His recommendation and what I went with was `CustomPipeName`. I've updated that throughout.

TylerLeonhardt committed:
[feature] touch up summary

iSazonov commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -129,25 +140,34 @@ protected override void EndProcessing()
                 case ProcessIdParameterSet:
                     Process = GetProcessById(Id);
+                    VerifyProcess(Process);
+                    namedPipeRunspace = CreateNamedPipeRunspace(Process.Id, AppDomainName);
```
Interesting, do we really need the VerifyProcess() check (that is File.Exist())? Perhaps the runtime error is informative enough.

TylerLeonhardt replied:
I'm not quite sure what you're saying. This `VerifyProcess` is to keep the behavior consistent with the current behavior.

iSazonov commented:
@TylerLeonhardt Please create new Issue in PowerShell-Docs repo and add reference to the PR description.

TylerLeonhardt committed:
move API tests to xUnit

TylerLeonhardt committed:
[feature] Add ParallelXUnitResults.xml

TylerLeonhardt committed:
[feature] remove debug log

TylerLeonhardt committed:
[feature] switch to Start-Process

TylerLeonhardt committed:
misc codacy fixes

TylerLeonhardt commented:
Alright CI is green and codacy and Code Factor are mostly happy. This should be good to go.

iSazonov commented on src/System.Management.Automation/engine/remoting/commands/EnterPSHostProcessCommand.cs:
```diff
@@ -345,6 +396,22 @@ private void VerifyProcess(Process process)
+        private void VerifyPipeName(string customPipeName)
+        {
+            // Named Pipes are represented differently on Windows vs macOS & Linux
+            var pipePath = Platform.IsWindows ? $@"\\.\pipe\{customPipeName}" : Path.Combine(Path.GetTempPath(), $"CoreFxPipe_{customPipeName}");
```
I don't like interpolate operator because it implicitly uses CurrentCulture but we always prefer set explicit value. Could we please use `string.Format`?

TylerLeonhardt replied:
moved to StringBuilder.

iSazonov reviewed (approved):
LGTM with minor comments.

TylerLeonhardt committed:
[feature] address Ilya's comments

TylerLeonhardt commented on test/xUnit/csharp/test_NamedPipe.cs:
```diff
@@ -0,0 +1,58 @@
+            Assert.True(File.Exists(GetPipePath(pipeName2)));
+
+            // Previous pipe should have been cleaned up.
+            Assert.True(!File.Exists(GetPipePath(pipeName1)));
```
fixed.

TylerLeonhardt commented on test/xUnit/csharp/test_NamedPipe.cs:
```diff
@@ -0,0 +1,58 @@
+        public void TestCustomPipeNameCreation()
+        {
+            string pipeName1 = Path.GetRandomFileName();
+            string pipeName2 = Path.GetRandomFileName();
```
fixed.

iSazonov commented:
@TylerLeonhardt Please create new Issue in PowerShell-Docs repo before we merge.

TylerLeonhardt commented:
Done, @iSazonov!

iSazonov commented:
@TylerLeonhardt Thanks for your contribution!
